src/ophyd_async/epics/adcore/_hdf_writer.py

```diff
@@ -90,16 +93,25 @@ async def open(self, multiplier: int = 1) -> Dict[str, DataKey]:
             )
         ]
         # And all the scalar datasets
-        for ds_name, ds_path in self._scalar_datasets_paths.items():
-            self._datasets.append(
-                HDFDataset(
-                    f"{name}-{ds_name}",
-                    f"/entry/instrument/NDAttributes/{ds_path}",
-                    (),
-                    "",
-                    multiplier,
-                )
-            )
+        try:
+            for plugin in self._plugins:
+                tree = ET.parse((await plugin.nd_attributes_file.get_value()))
+                root = tree.getroot()
+                for child in root:
+                    datakey = child.attrib["name"]
+                    self._datasets.append(
+                        HDFDataset(
+                            datakey,
+                            f"/entry/instrument/NDAttributes/{datakey}",
+                            (),
+                            convert_ad_dtype_to_np(
+                                ADBaseDataType((child.attrib.get("datatype", None)))
+                            ),
```

Comment:
This is not quite right, so I went reading the areaDetector code. Turns out to be more complicated than I thought.

There are 3 cases to consider:
- NDAttributeParam with type [DOUBLE, INT, INT64, STRING](https://github.com/areaDetector/ADCore/blob/1a9135976b4bc75bfdfe1ed7516a3e7d29e08e6b/ADApp/ADSrc/paramAttribute.cpp#L80-L95)
- NDAttributePV with type [DBR_NATIVE](https://github.com/areaDetector/ADCore/blob/1a9135976b4bc75bfdfe1ed7516a3e7d29e08e6b/ADApp/ADSrc/PVAttribute.cpp#L223-L256)
- NDAttributePV with type [DBR_LONG, DBR_STRING, ...](https://github.com/areaDetector/ADCore/blob/1a9135976b4bc75bfdfe1ed7516a3e7d29e08e6b/ADApp/ADSrc/PVAttribute.cpp#L257-L287)

I think we can solve with the following lookup:
```python
_ndattribute_to_ad_datatype = {
    "FLOAT": ADBaseDataType.Float64,
    "DBR_DOUBLE: ADBaseDataType.Float64,
    ...
}
if datatype in ["STRING", "DBR_STRING"]:
    np_datatype = "s40"
elif datatype == "DBR_NATIVE":
    raise ValueError("Don't support DBR_NATIVE yet")
else:
    ad_datatype = _ndattribute_to_ad_datatype[datatype]
    np_datatype = convert_ad_dtype_to_np(datatype)
```

Reply:
I used this comment to figure out the datatype https://github.com/areaDetector/ADCore/blob/1a9135976b4bc75bfdfe1ed7516a3e7d29e08e6b/ADApp/ADSrc/asynNDArrayDriver.cpp#L263